BL-155 Tenant isolation via policy

piotr-dulikowski opened the pull request:
**Work in progress**. Needs cleanup, doc update and - optionally - attachment policy checking when target resource is modified.

morrisson commented on tests/test_schema.yaml:
```diff
@@ -370,6 +370,42 @@ policies:
+  effect: allow
+  principal: Member
+  source:
+    resource_schema_id: attacher
```
source schema_id can be specified by resource.path (same policy can be applied to multiple types), so I do not think we need to add this property.

I will not stick to above idea, but it might be worth to consider.

piotr-dulikowski replied:
I removed this field.

morrisson commented on schema/policy.go:
```diff
@@ -260,12 +277,73 @@ func NewPolicy(raw interface{}) (*Policy, error) {
+	// TODO: Move validation somewhere else?
+	if policy.Action == ActionAttach {
+		// Source property is required
+		source, hasSource := typeData["source"].(map[string]interface{})
```
Maybe it's better to convert to struct.

marcin-ptaszynski commented on docs/policy.md:
```diff
@@ -263,3 +263,58 @@ In an example below admin has access to all methods in all paths except `delete`
+    - "allow": policy denies if source or target conditions are false
+    - "deny": policy denies if source and target conditions are true
+- resource : Like in the case of a standard policy, but considers the source resource
+- relation_property: Name of the relation property the policy is applied to
```
can we wildcard this value? Would be nice to have some defaults.

piotr-dulikowski replied:
Added possibility to specify a wildcard `relation_property`.

piotr-dulikowski commented:
This version is ready for review. Summary of the changes:
- Removed the "schema_id" field
- Removed the source condition field (its meaning was unclear to me, and don't think it is necessary for implementing tenant isolation)
- Updated documentation

marcin-ptaszynski commented on schema/policy.go:
```diff
@@ -34,6 +35,8 @@ const (
 	// ActionDelete allows to delete a resource
 	ActionDelete = "delete"
+	// ActionAttach allows a resource to have a relation to another resource
+	ActionAttach = "attach"
```
I'm afraid of potential custom action conflict - could we \_\_attach\_\_ instead?

piotr-dulikowski replied:
Changed "attach" to "__attach__".

morrisson commented on server/resources/resource_management.go:
```diff
@@ -1013,6 +1069,86 @@ func ActionResource(context middleware.Context, dataStore db.DB, identityService
+		if policy.IsDeny() {
+			return nil
+		} else {
+			return errorNotFound()
```
No related resource means user's input is incorrect. Therefore, `BadRequest()` should be better.

p-kozlowski replied:
changed, also modified the error construction, so that the callstack will point to the correct code path

morrisson commented on server/resources/resource_management.go:
```diff
@@ -1013,6 +1069,86 @@ func ActionResource(context middleware.Context, dataStore db.DB, identityService
+	err = otherCond.ApplyPropertyConditionFilter(schema.ActionAttach, relatedRes.Data(), nil)
+	if policy.IsDeny() {
+		if err == nil {
+			return errorNotFound()
```
Resource is available but the request is forbidden due to the policy. Therefore, `Forbidden()` should be better. (However, `BadRequest()` might be safer because error code which is different from above may imply the resource existence.)

p-kozlowski replied:
let's go the secure way. changed to `BadRequest`

p-kozlowski commented on docs/policy.md:
```diff
@@ -263,3 +263,59 @@ In an example below admin has access to all methods in all paths except `delete`
+The syntax is slightly different than in the case of a standard policy:
+
+- id, principal, resource, tenant_id: as in the case of a standard policy
+- action: Must be "__attach__"
```
fixed, same in other places

p-kozlowski commented:
@morrisson, remarks applied, squashed & rebased. please merge at your convenience

morrisson commented on docs/policy.md:
```diff
@@ -263,3 +263,59 @@ In an example below admin has access to all methods in all paths except `delete`
+The syntax is slightly different than in the case of a standard policy:
+
+- id, principal, resource, tenant_id: as in the case of a standard policy
+- action: Must be `attach`
```
nit: Must be `__attach__`
